### ISSUE TITLE
upstream connect error or disconnect/reset before headers. reset reason: local reset, transport failure reason: TLS error: 268435612:SSL routines:OPENSSL_internal:HTTP_REQUEST

### ISSUE THREAD

**Issue by lemonit-eric-mao** [istio/istio#32523]
(**NOTE**: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>)

**Bug description**

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[X] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

**Steps to reproduce the bug**

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)
``` ruby
[root@master01 ~]# istioctl version --remote
client version: 1.9.3
control plane version: 1.9.3
data plane version: 1.9.3 (3 proxies)


[root@master01 ~]# kubectl version --short
Client Version: v1.20.4
Server Version: v1.20.4

```

**How was Istio installed?**

**Environment where the bug was observed (cloud vendor, OS, etc)**

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

---

---

---

##### **[Deployment Istio operator](https://istio.io/latest/en/docs/setup/install/operator/ "Deployment Istio operator")**
`istioctl operator init`
``` ruby
istioctl --context="CTX_CLUSTER1" operator init

istioctl --context="CTX_CLUSTER2" operator init

```

---

---

---

###### Use the certificate provided by istio by default
``` ruby
kubectl --context="CTX_CLUSTER1" create ns istio-system

kubectl --context="CTX_CLUSTER1" create secret generic cacerts -n istio-system \
        --from-file=istio-1.9.3/samples/certs/ca-cert.pem \
        --from-file=istio-1.9.3/samples/certs/ca-key.pem \
        --from-file=istio-1.9.3/samples/certs/root-cert.pem \
        --from-file=istio-1.9.3/samples/certs/cert-chain.pem


kubectl --context="CTX_CLUSTER2" create ns istio-system

kubectl --context="CTX_CLUSTER2" create secret generic cacerts -n istio-system \
        --from-file=istio-1.9.3/samples/certs/ca-cert.pem \
        --from-file=istio-1.9.3/samples/certs/ca-key.pem \
        --from-file=istio-1.9.3/samples/certs/root-cert.pem \
        --from-file=istio-1.9.3/samples/certs/cert-chain.pem

```

---

---

---

###### Create an Istio configuration file for cluster1
``` ruby
cat> cluster1.yaml << ERIC
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  values:
    global:
      meshID: mesh1
      multiCluster:
        clusterName: cluster1
      network: network1
ERIC

```

##### Apply the configuration file to cluster1
``` ruby
istioctl --context="CTX_CLUSTER1" install -f cluster1.yaml

kubectl --context="CTX_CLUSTER1" get svc,po -n istio-system
```

---

###### By checking the operator controller log, you can detect the changes made by the controller in the cluster in response to the update of the IstioOperator CR
``` ruby
kubectl logs -f -n istio-operator $(kubectl get pods -n istio-operator -lname=istio-operator -o jsonpath='{.items[0].metadata.name}')

```

---

---

---

###### **[Official Document Multi-Cluster Installation](https://istio.io/latest/zh/docs/setup/install/multicluster/ "Official Document Multi-Cluster Installation")**

---

###### Create an Istio configuration file for cluster2
``` ruby
cat> cluster2.yaml << ERIC
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  values:
    global:
      meshID: mesh1
      multiCluster:
        clusterName: cluster2
      network: network1
ERIC

```

##### Apply the configuration file to cluster2
``` ruby
istioctl --context="CTX_CLUSTER2" install -f cluster2.yaml

kubectl --context="CTX_CLUSTER2" get svc,po -n istio-system
```

---

---

---

##### Turn on endpoint discovery, no need to switch k8s context, the following command will automatically select the context for creation
``` ruby
## Install the secret of the remote cluster in cluster2, which provides access to the API server of the cluster1 cluster.
[root@master01 ~]# istioctl x create-remote-secret \
    --context="CTX_CLUSTER1" \
    --name=cluster1 | \
    kubectl apply -f---context="CTX_CLUSTER2"

## View
[root@master01 ~]# kubectl get secret -n istio-system --context="CTX_CLUSTER2" | grep istio-remote
istio-remote-secret-cluster1 Opaque 1 7s


## ------------------------------------------------ -------------------------------------------------


## Install the secret of the remote cluster in cluster1, which provides access to the API server of cluster2.
[root@master01 ~]# istioctl x create-remote-secret \
    --context="CTX_CLUSTER2" \
    --name=cluster2 | \
    kubectl apply -f---context="CTX_CLUSTER1"

## View
[root@master01 ~]# kubectl get secret -n istio-system --context="CTX_CLUSTER1" | grep istio-remote
istio-remote-secret-cluster2 Opaque 1 5s

```

---

---

---

---

---

---

---

---

---

###### First, create a namespace sample in each cluster
``` ruby
kubectl create --context="CTX_CLUSTER1" namespace sample
kubectl create --context="CTX_CLUSTER2" namespace sample

```

---

###### Enable automatic sidecar injection for the namespace sample
``` ruby
kubectl label --context="CTX_CLUSTER1" namespace sample istio-injection=enabled
kubectl label --context="CTX_CLUSTER2" namespace sample istio-injection=enabled

```

---

###### Create HelloWorld `Service` in each cluster
``` ruby
kubectl apply --context="CTX_CLUSTER1" \
    -f istio-1.9.3/samples/helloworld/helloworld.yaml \
    -l service=helloworld -n sample

kubectl apply --context="CTX_CLUSTER2" \
    -f istio-1.9.3/samples/helloworld/helloworld.yaml \
    -l service=helloworld -n sample

```

---

---

---

###### Deploy the V1 version of HelloWorld
``` ruby
kubectl apply --context="CTX_CLUSTER1" \
    -f istio-1.9.3/samples/helloworld/helloworld.yaml \
    -l version=v1 -n sample

```

---

###### Deploy the V2 version of HelloWorld
``` ruby
kubectl apply --context="CTX_CLUSTER2" \
    -f istio-1.9.3/samples/helloworld/helloworld.yaml \
    -l version=v2 -n sample

```

---

---

---

###### Deploy Sleep
``` ruby
kubectl apply --context="CTX_CLUSTER1" \
    -f istio-1.9.3/samples/sleep/sleep.yaml -n sample

kubectl apply --context="CTX_CLUSTER2" \
    -f istio-1.9.3/samples/sleep/sleep.yaml -n sample

```

---

---

---

###### Verify cross-cluster traffic
`Send a request from the Sleep pod in cluster1 to the service HelloWorld`
``` ruby
kubectl exec --context="CTX_CLUSTER1" -n sample -c sleep \
    "$(kubectl get pod --context="CTX_CLUSTER1" -n sample -l \
    app=sleep -o jsonpath='{.items[0].metadata.name}')" \
    - curl helloworld.sample:5000/hello
```

---

###### Repeat this process with the Sleep pod in cluster2
``` ruby
kubectl exec --context="CTX_CLUSTER2" -n sample -c sleep \
     "$(kubectl get pod --context="CTX_CLUSTER2" -n sample -l \
     app=sleep -o jsonpath='{.items[0].metadata.name}')" \
     - curl helloworld.sample:5000/hello

```

---

---

---

##### Test
``` ruby
[root@master01 ~]# kubectl exec --context="CTX_CLUSTER1" -n sample -c sleep \
>     "$(kubectl get pod --context="CTX_CLUSTER1" -n sample -l \
>     app=sleep -o jsonpath='{.items[0].metadata.name}')" \
>     -- curl helloworld.sample:5000/hello
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    60  100    60    0     0    431      0 --:--:-- --:--:-- --:--:--   434
Hello version: v1, instance: helloworld-v1-776f57d5f6-zgbmz
[root@master01 ~]#
[root@master01 ~]#
[root@master01 ~]# kubectl exec --context="CTX_CLUSTER1" -n sample -c sleep \
>     "$(kubectl get pod --context="CTX_CLUSTER1" -n sample -l \
>     app=sleep -o jsonpath='{.items[0].metadata.name}')" \
>     -- curl helloworld.sample:5000/hello
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:09 --:--:--     0upstream connect error or disconnect/reset before headers. reset reason: local reset, transport failure reason: TLS100   175  100   175    0     0     17      0  0:00:10  0:00:10 --:--:--    46
[root@master01 ~]#

```

``` ruby
[root@master01 ~]# kubectl exec --context="CTX_CLUSTER2" -n sample -c sleep \
>     "$(kubectl get pod --context="CTX_CLUSTER2" -n sample -l \
>     app=sleep -o jsonpath='{.items[0].metadata.name}')" \
>     -- curl helloworld.sample:5000/hello
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0Hello version: v2, instance: helloworld-v2-54df5f84b-zg8vg
100    59  100    59    0     0    310      0 --:--:-- --:--:-- --:--:--   310
[root@master01 ~]#
[root@master01 ~]#
[root@master01 ~]# kubectl exec --context="CTX_CLUSTER2" -n sample -c sleep \
>     "$(kubectl get pod --context="CTX_CLUSTER2" -n sample -l \
>     app=sleep -o jsonpath='{.items[0].metadata.name}')" \
>     -- curl helloworld.sample:5000/hello
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   175  100   175    0     0     17      0  0:00:10  0:00:10 --:--:--    45
upstream connect error or disconnect/reset before headers. reset reason: local reset, transport failure reason: TLS error: 268435612:SSL routines:OPENSSL_internal:HTTP_REQUEST[root@master01 ~]#
[root@master01 ~]#

```

---

---

---

**Comment by crodriguezconde** [comment]
Experiencing the same issue in Istio 1.9.3 on GKE. When running https://istio.io/latest/docs/setup/install/multicluster/multi-primary/ I am receiving the same error message:

`upstream connect error or disconnect/reset before headers. reset reason: local reset, transport failure reason: TLS error: 268435612:SSL routines:OPENSSL_internal:HTTP_REQUEST`

Is there any prerequisites on cloud providers such as GCP to make this work?

**Comment by aasier** [comment]
Hi all!
Same issue deploying istio with the version :

![image](https://user-images.githubusercontent.com/17740475/118264973-d9715f00-b4b8-11eb-9403-dfb14d0829e6.png)

Provider Azure AKS.

**Comment by ignatenko451** [comment]
Hello. Have the same problem in Istio 1.9.5 on bare-metal k8s

**Comment by dimashenkov** [comment]
Hi guys. I have two clusters with istio 1.9.0 . When I configure Multi-Primary multicluster ,after execution of command istioctl x create-remote-secret,some of my apps get Upstream connect error or disconnect/reset before headers.Any idea why?

**Comment by nmittler** [comment]
@stevenctl mind taking a look?

**Comment by stevenctl** [comment]
I'll try this out myself for the latest 1.9.x release – make sure to try out the steps at https://istio.io/latest/docs/ops/diagnostic-tools/multicluster/

**Comment by ignatenko451** [comment]
Hello. Multicluster's troubleshoot didn't get any conflicts with recommended load balancing, trust and proxy configurations. 
As before, connection error, which described above, randomly occurs when I try to send request from sleep pod to helloworld pod.
~50% requests end with error, other requests get responces, but only in those cluster, from which I send requests (cross-cluster load balancing does not work)

**Comment by FeiYing9** [comment]
> 
> 
> I am using vm sidecar，hava the same error？
> upstream connect error or disconnect/reset before headers. reset reason: local reset, transport failure reason: TLS error: 268435612:SSL routines:OPENSSL_internal:HTTP_REQUEST

Mee too.    with the 1.9.2 of isito and 1.18.2 of k8s.

**Comment by stevenctl** [comment]
The most common cause of something like this is that the two clusters are not using a common root of trust. I haven't been able to replicate this. Are there any other certs/trust related settings enabled that deviate from the simple example in the docs?

**Comment by Shabirmean** [comment]
@stevenctl Thank you for looking into this. I have faced this same issue/error with the [`Hello World` example](https://cloud.google.com/service-mesh/docs/gke-install-multi-cluster) for ASM on GKE which I believe is using Istio.  The steps don't seem to involve any changes to CAs/Certs etc

**Comment by stevenctl** [comment]
To everyone in the thread, did all of you follow the instructions at https://istio.io/latest/docs/setup/install/multicluster/before-you-begin/#configure-trust? 

We intend for the steps under https://istio.io/latest/docs/tasks/security/cert-management/plugin-ca-cert/#plug-in-certificates-and-key-into-the-cluster to be followed once for each of your clusters. 

The way the docs are laid out is a bit confusing, so this step is easy to miss. This might be worth changing if it's too unclear.

**Comment by stevenctl** [comment]
cc @Shabirmean  @FeiYing9 @ignatenko451 @dimashenkov @13567436138 @crodriguezconde 

It does seem @lemonit-eric-mao followed these steps, and I recommend double checking it was done correctly (see https://istio.io/latest/docs/ops/diagnostic-tools/multicluster/#trust-configuration)

**Comment by chadlwilson** [comment]
Could this be due to the same root issue as https://github.com/istio/istio/issues/33105 (restricted TLS ciphers in Envoy, and client/server cannot agree a TLS 1.2 cipher)?

If someone is able to packet capture traffic using something like ksniff and see the server TCP RST the client after TLS CLIENT HELLO, it could be.

**Comment by dimashenkov** [comment]
I have the same err , and also after execution of command `istioctl x create-remote-secret`,some of my apps get `Upstream connect error or disconnect/reset before headers` from the remote cluster. Any idea why?

**Comment by Shabirmean** [comment]
@stevenctl - Thank you for those useful links. I have tried the troubleshooting links and the issue still seems to persist.

Also noteworthy - I get the issue with managed control plane using istio in GCP Anthos Service Mesh (ASM). It has no CA configurations and uses MeshCA. So I didn't have to configure any CA stuff

**Comment by stevenctl** [comment]
@Shabirmean @FeiYing9 @ignatenko451 @dimashenkov @13567436138 @crodriguezconde

Can you share your vendor and kubernetes version? Trying to determine if this is specific to 1.20 and/or GKE

**Comment by Shabirmean** [comment]
> @Shabirmean @FeiYing9 @ignatenko451 @dimashenkov @13567436138 @crodriguezconde
> 
> Can you share your vendor and kubernetes version? Trying to determine if this is specific to 1.20 and/or GKE

@stevenctl - I have trying this on GKE Version **1.19.9-gke.1400** / ASM v1-9-5

**Comment by Shabirmean** [comment]
The issue related to this not working in GKE with managed Istio was due to default firewall rules on the GCP Project. Was able to get the [Hello World example](https://cloud.google.com/service-mesh/docs/gke-install-multi-cluster) + the others working by adding the required firewall rules

```
ALL_CLUSTER_CIDRS=$(gcloud container clusters list --format='value(clusterIpv4Cidr)' | sort | uniq)
ALL_CLUSTER_CIDRS=$(join_by , $(echo "${ALL_CLUSTER_CIDRS}"))
ALL_CLUSTER_NETTAGS=$(gcloud compute instances list --format='value(tags.items.[0])' | sort | uniq)
ALL_CLUSTER_NETTAGS=$(join_by , $(echo "${ALL_CLUSTER_NETTAGS}"))

gcloud compute firewall-rules create istio-multicluster-test-pods \
  --allow=tcp,udp,icmp,esp,ah,sctp \
  --direction=INGRESS \
  --priority=900 \
  --source-ranges="${ALL_CLUSTER_CIDRS}" \
  --target-tags="${ALL_CLUSTER_NETTAGS}" --quiet
```

The same on the istio docs side has been updated in this [istio docs PR](https://github.com/istio/istio.io/pull/9942)

**Comment by dimashenkov** [comment]
> @Shabirmean @FeiYing9 @ignatenko451 @dimashenkov @13567436138 @crodriguezconde
> 
> Can you share your vendor and kubernetes version? Trying to determine if this is specific to 1.20 and/or GKE

AWS EKS k8s v.1.20

**Comment by Shabirmean** [comment]
@dimashenkov - Mine was **_GKE 1.19.10-gke.1600_**

**Comment by stevenctl** [comment]
For now my assumption is that most users facing this have a general connectivity issue. Can the other users facing this try the test in this comment https://github.com/istio/istio/issues/32523#issuecomment-857250510.

**Comment by dimashenkov** [comment]
> For now my assumption is that most users facing this have a general connectivity issue. Can the other users facing this try the test in this comment [#32523 (comment)](https://github.com/istio/istio/issues/32523#issuecomment-857250510).

`* Rebuilt URL to: 10.172.180.28/
*   Trying 10.172.180.28...
* TCP_NODELAY set



* connect to 10.172.180.28 port 80 failed: Connection timed out
* Failed to connect to 10.172.180.28 port 80: Connection timed out
* Closing connection 0
curl: (7) Failed to connect to 10.172.180.28 port 80: Connection timed out
command terminated with exit code 7`

**Comment by stevenctl** [comment]
@dimashenkov, the helloworld app runs on port 5000, can you try again with the proper port?

**Comment by dimashenkov** [comment]
@stevenctl 
`  -- sh -c 'curl -v 10.172.178.162:5000'
* Rebuilt URL to: 10.172.178.162:5000/
*   Trying 10.172.178.162...
* TCP_NODELAY set
* Connected to 10.172.178.162 (10.172.178.162) port 5000 (#0)
> GET / HTTP/1.1
> Host: 10.172.178.162:5000
> User-Agent: curl/7.58.0
> Accept: */*
>
* HTTP 1.0, assume close after body
< HTTP/1.0 404 NOT FOUND
< Content-Type: text/html
< Content-Length: 233
< Server: Werkzeug/0.12.2 Python/2.7.13
< Date: Thu, 01 Jul 2021 16:18:55 GMT
<
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>404 Not Found</title>
<h1>Not Found</h1>
<p>The requested URL was not found on the server.  If you entered the URL manually please check your spelling and try again.</p>
* Closing connection 0`

**Comment by dimashenkov** [comment]
@stevenctl 
looks like there are connectivity but still HelloWorld version didnt toggle between v1 and v2 ,when I test cross-cluster load balancing

**Comment by stevenctl** [comment]
Thanks for testing that. 

In https://github.com/istio/istio/issues/32523#issuecomment-850221366 you said "they are not even in the istio-system namespace", are you talking about your apps that are having issues? I'm not sure what the "two problems" are here.

**Comment by dimashenkov** [comment]
@stevenctl 
1.after execution of command istioctl x create-remote-secret, it breaks some of my other apps get Upstream connect error or disconnect/reset before headers
2. HelloWorld version didnt toggle between v1 and v2 ,when I test cross-cluster load balancing

**Comment by stevenctl** [comment]
I know this was a while ago, but I assume `10.172.180.28` was the Pod IP of an instance on the _remote_ cluster?

Perhaps we should open a new issue.  The conversation here to me can be summarized as: "Many users saw a TLS error when in fact there was actually a general connectivity problem". Yours may be different.

**Comment by dimashenkov** [comment]
> I know this was a while ago, but I assume `10.172.180.28` was the Pod IP of an instance on the _remote_ cluster?
> 
> Perhaps we should open a new issue. The conversation here to me can be summarized as: "Many users saw a TLS error when in fact there was actually a general connectivity problem". Yours may be different.

I had TLS too, I opened a new issue but nobody answer https://github.com/istio/istio/issues/33848

I think all this problems are connected

**Comment by saumilsdk** [comment]
Here is how I fixed it for KF 1.4 and K8 1.21

![image](https://user-images.githubusercontent.com/8816702/176830699-7d797cf2-86c0-42f1-bf20-e1366ea72969.png)
